### PR TITLE
Added a functionality to log results from a terraform command into a …

### PR DESCRIPTION
**Pull request by Doreen9**
##### SUMMARY
Added the functionality to log activities (example performing a terraform apply, run or destroy, outcomes(resources added or destroyed or changed)) in a json format into an audit file.

format:
`  {
         'time': timestamp,
         'plan_used': plan_used,
         'command_run': command_used,
         'outcome': outcome,
         'resources_added': list_creation,
         'resources_changed': list_modification,
         'resources_destroyed': list_destruction
         }`

The current module creates a temporary plan files every time you run the 'apply' command or a 'plan' command  if you do not provide a path to a plan_file. 

The contents of the audit file helps to retrieve the path to last plan file generated automatically by the module when the 'plan' command was run so that the user has the option to reuse the plan file generated by providing an id that would be displayed to the user upon running the 'plan' command without providing a path to the plan file.

Also added test cases for some already existing functions in the module.

Fixes to [#5 ](https://github.com/turntabl/community.general/issues/5) and [#3](https://github.com/turntabl/community.general/issues/3)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
terraform module

##### ADDITIONAL INFORMATION

The model currently takes additional parameters `plan_id` provided by the user in case they might wish to reuse a plan file, `log_activity_folder`  path to a folder to containing the file to log the activities as stated earlier, and `file_name` to create the audit file and name the file with. 

If the user does not provide the `plan_id` and a `plan_path` and he or she performs a plan before an apply, the module retrieves the path to the last plan file generated by the user from the audit file and uses that plan file.
 
The already existing functions in the module did not have test cases so a few test cases have been added.

example:

playbook.yml:
```
- name: terraform 
    terraform:
      project_path: '/tmp/project/'
      plan_id: 'tmpER44'
      log_activity_folder: '/home/user/'
      file_name: 'jnnj4444'
      state: present
      check_deletes: True

```

when this playbook is executed it will create a file named `jnnj4444` in the `/home/user/` directory.
contents of `jnnj4444`:
`
[
    {
        "plan_used": "/tmp/tmpER44.tfplan",
        "resources_added": [],
        "resources_destroyed": [],
        "time": "21-05-2020 23:51:08",
        "command_run": "apply",
        "outcome": "Apply complete! Resources: 0 added, 0 changed, 0 destroyed."   
        "resources_changed": []
     }
]
`

### CODE REVIEW

File: changelogs/fragments/574-logging-terraform-outcomes.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - terraform - extra parameters(plan_id, log_activity_folder, file_name) are added to log outcomes from
```
Comment (felixfontein) [inline changelogs/fragments/574-logging-terraform-outcomes.yaml:2]:
```suggestion
  - terraform - extra parameters (``plan_id``, ``log_activity_folder``, ``file_name``) are added to log outcomes from 
```

File: changelogs/fragments/574-logging-terraform-outcomes.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - terraform - extra parameters(plan_id, log_activity_folder, file_name) are added to log outcomes from 
+  a terraform command. (https://github.com/ansible-collections/community.general/pull/574).
```
Comment (felixfontein) [inline changelogs/fragments/574-logging-terraform-outcomes.yaml:3]:
```suggestion
    a terraform command. (https://github.com/ansible-collections/community.general/pull/574).
```

File: plugins/modules/cloud/misc/terraform.py [hunk 3]
```diff
@@ -93,6 +93,24 @@
     type: list
     elements: path
     version_added: '0.2.0'
+  plan_id:
+    description:
+      - A plan_id linked to a temporary plan file created if no plan_path
+        is specified.
+    required: false
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/terraform.py:100]: You can leave `required: false` away (it's the default). But what you must do is add `type: xxx`, where `xxx` is replaced by the correct type.

File: plugins/modules/cloud/misc/terraform.py [hunk 4]
```diff
@@ -287,8 +408,10 @@ def main():
             force_init=dict(type='bool', default=False),
             backend_config=dict(type='dict', default=None),
             backend_config_files=dict(type='list', elements='path', default=None),
+            plan_id=dict(type='str'),
+            log_activity_folder=dict(type='path', default='./'),
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/terraform.py:412]: `default` must also be mentioned in the documentation.

File: tests/unit/plugins/modules/cloud/misc/test_terraform.py [hunk 5]
```diff
@@ -1,12 +1,188 @@
-# Copyright: (c) 2019, Ansible Project
```
Comment (felixfontein) [inline tests/unit/plugins/modules/cloud/misc/test_terraform.py:1]: Please don't touch copyright headers of existing files.

File: tests/unit/plugins/modules/cloud/misc/test_terraform.py [hunk 6]
```diff
@@ -1,12 +1,188 @@
...
-# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
-
```
Comment (felixfontein) [inline tests/unit/plugins/modules/cloud/misc/test_terraform.py:3]: This file was modified in another PR (some common boilerplate code was added to many different files in there), you need to rebase your PR.

File: plugins/modules/cloud/misc/terraform.py [hunk 7]
```diff
@@ -287,8 +408,10 @@ def main():
...
+            file_name=dict(type='str', default='audit'),
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/terraform.py:413]: Does the default here mean that logging is enabled by default? That would be a breaking change. I'd rather make it optional, but not the new default.

File: tests/unit/plugins/modules/cloud/misc/test_terraform.py (hunk 6 above)
Comment (Doreen9) [reply tests/unit/plugins/modules/cloud/misc/test_terraform.py:3]: sure. will do that

File: tests/unit/plugins/modules/cloud/misc/test_terraform.py (hunk 5 above)
Comment (Doreen9) [reply tests/unit/plugins/modules/cloud/misc/test_terraform.py:1]: oops! My bad.

File: plugins/modules/cloud/misc/terraform.py (hunk 7 above)
Comment (Doreen9) [reply plugins/modules/cloud/misc/terraform.py:413]: please yes. It means that logging is enabled by default. I will make it optional instead.

**Pull request closed without merging**